From ec0f2d7c8198f368d97ef070186c1b3b5a78bbea Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Wed, 20 May 2015 14:59:23 +0800
Subject: [PATCH] MLK-10930-3 usb: chipidea: add delay if phy-clkgate-delay-us
 property is present

For some platforms, time delay is requried between putting PHY into low power
mode and gate PHY clock.

Signed-off-by: Li Jun <jun.li@freescale.com>
---
 drivers/usb/chipidea/core.c  | 7 +++++++
 include/linux/usb/chipidea.h | 1 +
 2 files changed, 8 insertions(+)

diff --git a/drivers/usb/chipidea/core.c b/drivers/usb/chipidea/core.c
index b096946..8a91b5a 100644
--- a/drivers/usb/chipidea/core.c
+++ b/drivers/usb/chipidea/core.c
@@ -626,6 +626,10 @@ static int ci_get_platdata(struct device *dev,
 	if (of_usb_get_maximum_speed(dev->of_node) == USB_SPEED_FULL)
 		platdata->flags |= CI_HDRC_FORCE_FULLSPEED;
 
+	if (of_find_property(dev->of_node, "phy-clkgate-delay-us", NULL))
+		of_property_read_u32(dev->of_node, "phy-clkgate-delay-us",
+					&platdata->phy_clkgate_delay_us);
+
 	return 0;
 }
 
@@ -986,6 +990,9 @@ static void ci_controller_suspend(struct ci_hdrc *ci)
 {
 	disable_irq(ci->irq);
 	ci_hdrc_enter_lpm(ci, true);
+	if (ci->platdata->phy_clkgate_delay_us)
+		usleep_range(ci->platdata->phy_clkgate_delay_us,
+				ci->platdata->phy_clkgate_delay_us + 50);
 	usb_phy_set_suspend(ci->usb_phy, 1);
 	ci->in_lpm = true;
 	enable_irq(ci->irq);
diff --git a/include/linux/usb/chipidea.h b/include/linux/usb/chipidea.h
index 97f0b5c..d1f8f55 100644
--- a/include/linux/usb/chipidea.h
+++ b/include/linux/usb/chipidea.h
@@ -49,6 +49,7 @@ struct ci_hdrc_platform_data {
 	bool			tpl_support;
 	u32			ahbburst_config;
 	u32			burst_length;
+	u32			phy_clkgate_delay_us;
 };
 
 /* Default offset of capability registers */
-- 
1.8.0

